BEIMA.Backend/MongoService/BaseDevice.cs

```diff
@@ -0,0 +1,174 @@
+            Location = new BsonDocument();
+            LastModified = new BsonDocument();
+        }
+
+        /// <summary>
+        /// Full constructor to instantiate a Device object.
+        /// </summary>
+        /// <param name="id">ObjectId of the object itself.</param>
+        /// <param name="deviceTypeId">ObjectId of the DeviceType this Device is linked to.</param>
+        /// <param name="deviceTag"></param>
+        /// <param name="manufacturer"></param>
+        /// <param name="modelNum"></param>
+        /// <param name="serialNum"></param>
+        /// <param name="yearManufactured"></param>
+        /// <param name="notes"></param>
+        public BaseDevice(ObjectId id, ObjectId deviceTypeId, string deviceTag, string manufacturer, string modelNum, string serialNum, int yearManufactured, string notes)
+        {
+            //BsonClassMap.RegisterClassMap<BaseDevice>();
+
+            Id = id;
+            DeviceTypeId = deviceTypeId;
+            DeviceTag = deviceTag ?? string.Empty;
+            Manufacturer = manufacturer ?? string.Empty;
+            ModelNum = modelNum ?? string.Empty;
+            SerialNum = serialNum ?? string.Empty;
+            YearManufactured = yearManufactured;
+            Notes = notes ?? string.Empty;
+            Fields = new BsonDocument();
+            Location = new BsonDocument();
+            LastModified = new BsonDocument();
+        }
+
+
+        [BsonId]
+        public ObjectId Id { get; set; }
+
+        [BsonElement("deviceTypeId")]
+        public ObjectId DeviceTypeId { get; set; }
+
+        [BsonElement("deviceTag")]
+        public string DeviceTag { get; set; }
+
+        [BsonElement("manufacturer")]
+        public string Manufacturer { get; set; }
+
+        [BsonElement("modelNum")]
+        public string ModelNum { get; set; }
+
+        [BsonElement("serialNum")]
+        public string SerialNum { get; set; }
+
+        [BsonElement("yearManufactured")]
+        public int YearManufactured { get; set; }
+
+        [BsonElement("notes")]
+        public string Notes { get; set; }
+
+        [BsonElement("fields")]
+        public BsonDocument Fields { get; set; }
+
+        [BsonElement("location")]
+        public BsonDocument Location { get; set; }
+
+        [BsonElement("lastModified")]
+        public BsonDocument LastModified { get; set; }
+
+        /// <summary>
+        /// Gets a BsonDocument that represents the current state of the BaseDevice object.
+        /// </summary>
+        /// <returns>BsonDocument that represents the current state of the BaseDevice object. Conforms to the schema located in BEIMA.DB.Schemas.</returns>
+        /// <exception cref="ArgumentNullException">Throws exception when any of the required fields are null.</exception>
+        public BsonDocument GetBsonDocument()
+        {
+            if (DeviceTag == null)
+            {
+                throw new ArgumentNullException($"BaseDevice - {nameof(DeviceTag)} is null");
```

Comment:
Not sure how I feel about throwing an exception here on null data, especially for things like Notes or LastModified that may not necessarily be required have any data in them. If we want to throw an exception, I think it would be better as a NullReferenceException instead of ArgumentNullException?

Reply:
I think we should only throw exceptions on data we think should absolutely be required.

Reply:
Null data indicates that the programmer did not pass anything in for that field. The programmer needs to pass in an empty string, or some other default value. If null data has been passed in, it means something has gone wrong and an exception must be thrown.

Reply:
Comparing the descriptions of both exceptions, it seemed like the wording of ArgumentNullException seems more appropriate to me.
https://docs.microsoft.com/en-us/dotnet/api/system.nullreferenceexception?view=net-6.0
https://docs.microsoft.com/en-us/dotnet/api/system.argumentnullexception?view=net-6.0

It seems like a NullReferenceException is thrown when "there is an attempt to dereference a null object reference."

Versus, a ArgumentNullException is thrown when "a null reference is passed to a method that does not accept it as a valid argument." This seems more appropriate, since the only way those values can be null, is if a null was passed in as an argument to the constructor/object initializer.